ForumProject/notifications/signals.py

```diff
@@ -39,47 +49,83 @@ def create_notifications(instances, follow_status_change=True):
+                print(message) # This is a placeholder for "in app" / push notifications implementation
+        else:
+            if startup.notice_preferences.email_on_share_subscription:
+                # Asynchronously send email
+                email_thread = threading.Thread(target=send_email_async, args=(subject, message, recipients))
+                email_thread.start()
+            if startup.notice_preferences.in_app_on_share_subscription:
+                print(message) # This is a placeholder for "in app" / push notifications implementation
+
+# Signal to record and send a notification when an investor starts following a project or changes a share
+@receiver(post_save, sender=InvestorProject)
+def project_followed_or_subscription_changed(sender, instance, created, **kwargs):
+    """
+    Signal handler to create Notification records and send email notifications when an 
+    InvestorProject is created or updated.
+
+    Args:
+    - sender (type): The model class that sent the signal.
+    - instance (InvestorProject): The InvestorProject instance that triggered the signal.
     - created (bool): Whether the `InvestorProject` instance was just created.
-    - kwargs (dict): Additional arguments for the signal handler.
     """
     if created:
         # Investor just started following a project (share is initialized to 0 or another value)
-        create_notifications(instance)
+        do_notifications(instance)
+        if instance.share:
+            do_notifications(instance, follow_status_change=False)
     else:
         # Investor changed the share (subscription)
-        create_notifications(instance, follow_status_change=False)
+        do_notifications(instance, follow_status_change=False)
 
-# Signal to create a notification when an investor delists a project
+# Signal to record and send a notification when an investor delists a project
 @receiver(post_delete, sender=InvestorProject)
-def investor_project_delisted(sender, instance, **kwargs):
+def project_delisted(sender, instance, **kwargs):
     """
-    Signal handler that creates a notification when an `InvestorProject` record is deleted.
-
-    This signal is triggered when an investor delists a project, indicating that they 
-    have stopped following the project. It creates a `Notification` record to represent 
-    this change.
+    Signal handler to create a Notification record and send email notifications when an InvestorProject is deleted.
 
-    Parameters:
-    - sender (type): The model class that sent the signal (usually `InvestorProject`).
-    - instance (InvestorProject): The instance that triggered the signal.
-    - kwargs (dict): Additional arguments for the signal handler.
+    Args:
+    - sender (type): The model class that sent the signal.
+    - instance (InvestorProject): The InvestorProject instance that triggered the signal.
     """
-    create_notifications(instance)
+    do_notifications(instance)
```

Comment:
Since the instance is deleted, the information needed to send notifications (like related `UserStartup` or email preferences) might not be available or relevant anymore. Consider the implications of sending notifications for deleted objects and ensure that all necessary data is still available.